In [1]:
import pandas as pd

In [2]:
df = pd.read_csv("hr.csv")

# Fase 1 pareja A

In [ ]:
# TAREA A1 — ¿Qué forma tiene el dataset?

# ── 1. Cuántas filas y columnas hay ──────────────
print("Filas    :", df.shape[0])
print("Columnas :", df.shape[1])

# ── 2. Qué tipo de dato tiene cada columna ───────
print("\nTIPOS DE DATOS:")
print(df.dtypes.to_string())

# ── 3. Ver las primeras 10 filas ─────────────────
df.head(10)

Qué hace cada línea:
df.shape[0] → número de filas → nos dice cuántos empleados hay en el dataset
df.shape[1] → número de columnas → nos dice cuántas variables tenemos de cada empleado
df.dtypes.to_string() → muestra el tipo de dato de cada columna:

int64 → número entero ✅
float64 → número decimal ⚠️ algunas deberían ser int
str → texto ✅

df.head(10) → muestra las primeras 10 filas para ver cómo son los datos a simple vista. Ponemos 10 en lugar de 5 para ver más ejemplos

### Hallazgos Tarea A1
- El dataset tiene **1474 empleados** y **35 variables**
- Hay columnas mal tipadas — son float64 pero deberían ser int:
    - `Age` → 41.0 debería ser 41
    - `MonthlyIncome` → debería ser int
    - `JobSatisfaction` → escala 1-4, no necesita decimales
    - `YearsWithCurrManager` → los años son enteros
- Esto se arreglará en la **Fase 2**

👀 Lo que nos dice este resultado:
Tipos mal tipados — float64 que deberían ser int:

Age → 41.0, 49.0, 37.0... → deberían ser 41, 49, 37
JobSatisfaction → escala del 1 al 4, sin decimales
MonthlyIncome → salario sin decimales
YearsWithCurrManager → años enteros
StandardHours → siempre 80.0 → además inútil 🗑️

Ya vemos NaN a simple vista en df.head(10):

Fila 1 → StandardHours = NaN
Fila 6 → Department = NaN y YearsWithCurrManager = NaN
Fila 7 → Department = NaN
Fila 9 → EducationField = NaN y YearsWithCurrManager = NaN

### Hallazgos Tarea A1
- El dataset tiene **1474 empleados** y **35 variables**
- Columnas mal tipadas (float64 → deberían ser int):
    - Age, JobSatisfaction, MonthlyIncome, YearsWithCurrManager
- Ya se ven NaN a simple vista en las primeras 10 filas
- Se arreglará todo en la Fase 2

In [ ]:
# TAREA A2 — ¿Hay huecos vacíos?

# ── 1. Contar nulos por columna ──────────────────
print("NULOS POR COLUMNA:")
nulos = df.isnull().sum()
print(nulos[nulos > 0].sort_values(ascending=False))

# ── 2. Porcentaje de nulos por columna ───────────
print("\nPORCENTAJE DE NULOS:")
porcentaje = df.isnull().mean() * 100
porcentaje = porcentaje[porcentaje > 0].round(2).sort_values(ascending=False)
print(porcentaje)

Qué hace cada línea:
df.isnull().sum() → cuenta cuántos nulos hay en cada columna

df.isnull() → crea una tabla de True/False — True donde hay nulo
.sum() → suma los True de cada columna → total de nulos

nulos[nulos > 0] → filtra y se queda solo con las columnas que tienen algún nulo — las que tienen 0 no nos interesan
.sort_values(ascending=False) → ordena de mayor a menor — las más problemáticas primero
df.isnull().mean() * 100 → en lugar de contar, calcula la proporción de nulos:

.mean() → calcula la media de True/False → da un número entre 0 y 1
* 100 → lo convierte en porcentaje

👀 Lo que nos dice este resultado:
11 columnas tienen nulos de las 35 totales. Organizados por gravedad:
🔴 Más del 10% — graves:

StandardHours → 164 nulos (11.13%) → además siempre vale 80 → eliminar en Fase 2 🗑️
YearsWithCurrManager → 148 nulos (10.04%) → rellenar con mediana

🟡 Entre 5% y 10% — moderados:

MaritalStatus → 132 nulos (8.96%) → además tiene el typo Marreid → rellenar con moda
BusinessTravel → 117 nulos (7.94%) → rellenar con moda
TrainingTimesLastYear → 88 nulos (5.97%) → rellenar con mediana
Age → 73 nulos (4.95%) → rellenar con mediana

🟢 Menos del 5% — leves:

EducationField → 58 nulos (3.93%) → rellenar con moda
OverTime → 44 nulos (2.99%) → rellenar con moda
Department → 29 nulos (1.97%) → rellenar con moda
JobSatisfaction → 29 nulos (1.97%) → rellenar con mediana
MonthlyIncome → 14 nulos (0.95%) → rellenar con mediana

### Hallazgos Tarea A2
- **11 columnas** tienen valores nulos de 35 totales
- La más grave: StandardHours con 164 nulos (11.13%)
  → además es constante (siempre vale 80) → eliminar en Fase 2
- Columnas numéricas con nulos → rellenar con **mediana** en Fase 2:
  YearsWithCurrManager, TrainingTimesLastYear, Age,
  JobSatisfaction, MonthlyIncome
- Columnas de texto con nulos → rellenar con **moda** en Fase 2:
  MaritalStatus, BusinessTravel, EducationField,
  OverTime, Department

In [ ]:
# TAREA A3 — ¿Hay filas repetidas?

# ── 1. Contar filas duplicadas ───────────────────
duplicados = df.duplicated().sum()
print("Filas duplicadas:", duplicados)

# ── 2. Ver cómo son esas filas duplicadas ────────
if duplicados > 0:
    print("\nAsí son las filas duplicadas:")
    display(df[df.duplicated(keep=False)])

Qué hace cada línea:
df.duplicated() → marca como True las filas exactamente iguales a otra anterior
.sum() → cuenta cuántas hay
if duplicados > 0: → solo ejecuta lo siguiente si hay duplicados — si no los hay no tiene sentido buscarlos
df.duplicated(keep=False) → esto es nuevo respecto a antes 😊

Sin keep=False → solo marca la segunda copia como duplicada
Con keep=False → marca todas las copias — así vemos tanto el original como la copia duplicada

display(df[df.duplicated(keep=False)]) → muestra todas las filas duplicadas en formato tabla para poder verlas

¿Qué nos dice este resultado?
Aparecen 8 filas pero son 4 pares — cada duplicado aparece dos veces (el original y la copia):
OriginalCopiaEmpleadoFila 67Fila 147345 años, R&D, 25 años trabajandoFila 184Fila 147153 años, R&D, MedicalFila 1041Fila 147028 años, Sales, MedicalFila 1222Fila 147224 años, HR, abandonó (Yes)

💡 Detalle importante — fíjate en los índices:
Las copias están en las filas 1470, 1471, 1472, 1473 — las últimas del dataset. Esto sugiere que alguien añadió datos duplicados al final del archivo por error 😊

### Hallazgos Tarea A3
- Hay **4 filas duplicadas** en el dataset
- Las copias están al final del archivo (filas 1470-1473)
  → probablemente error al añadir datos
- Se eliminarán en la **Fase 2** con df.drop_duplicates()
- Tras eliminarlas quedarán **1470 empleados** en lugar de 1474

In [ ]:
# TAREA A4 — Estadísticas de columnas numéricas

# ── 1. Estadísticas generales ────────────────────
print("ESTADISTICAS COLUMNAS NUMERICAS:")
display(df.describe().T)

# ── 2. Columnas más relevantes para el proyecto ──
print("\nCOLUMNAS CLAVE — detalle:")

columnas_clave = ["Age", "MonthlyIncome",
                  "YearsAtCompany", "TotalWorkingYears",
                  "JobSatisfaction", "WorkLifeBalance"]

for columna in columnas_clave:
    minimo  = df[columna].min()
    maximo  = df[columna].max()
    media   = df[columna].mean().round(2)
    mediana = df[columna].median()
    print(f"\n{columna}:")
    print(f"  Min: {minimo}  |  Max: {maximo}  |  Media: {media}  |  Mediana: {mediana}")

Qué hace cada línea:
df.describe().T → ya lo conoces — estadísticas de todas las numéricas. El .T transpone la tabla para que sea más legible 😊
columnas_clave = [...] → lista con las columnas más importantes para el proyecto — las que más nos interesan para entender la retención de empleados
for columna in columnas_clave: → recorre cada columna de la lista
df[columna].min() → el valor mínimo de esa columna
df[columna].max() → el valor máximo
df[columna].mean().round(2) → la media redondeada a 2 decimales
df[columna].median() → la mediana — el valor del medio
print(f"\n{columna}:") → imprime el nombre de la columna con un salto de línea antes para que quede separado visualmente

¡Salió todo perfectamente! ✅ Tarea A4 completada al 100%!

👀 Las columnas clave nos dicen:
ColumnaMinMaxMediaMedianaConclusiónAge186036.9436Plantilla joven y equilibrada ✅MonthlyIncome10091999964974907Media > mediana → hay salarios muy altos que tiran hacia arriba ⚠️YearsAtCompany04075La mitad lleva menos de 5 años 👀TotalWorkingYears04011.2710Experiencia media razonable ✅JobSatisfaction142.733Satisfacción media-baja ⚠️WorkLifeBalance142.763Equilibrio vida-trabajo medio ⚠️

💡 El dato más importante:
MonthlyIncome — media 6497 pero mediana 4907. Diferencia de casi $1,600. Esto significa que hay pocos empleados con salarios muy altos que suben la media. La mayoría cobra alrededor de $4,907 — la mediana es más honesta aquí 😊

### Hallazgos Tarea A4
- **Salario**: media $6,497 pero mediana $4,907
  → la media está inflada por salarios muy altos (outliers)
  → la mediana es más representativa
- **Edad**: entre 18 y 60, media 36.9 → plantilla joven
- **Años en empresa**: mediana 5 años → la mitad lleva menos
  de 5 años → plantilla relativamente nueva
- **Satisfacción laboral**: media 2.73/4 → tirando a baja
- **WorkLifeBalance**: media 2.76/4 → mejorable
- **EmployeeCount y StandardHours**: std=0 → inútiles
  → eliminar en Fase 2

🎊 ¡FASE 1 PAREJA A COMPLETADA! 🎉
Esto es lo que tienes listo para la puesta en común con la Pareja B:

Tarea                            Hallazgo principal
A1                            4 columnas mal tipadas (float → int)
A2                           11 columnas con nulos — StandardHours la peor (11.13%)
A3                           4 filas duplicadas — al final del archivo
A4                           Salario con outliers, satisfacción baja, plantilla joven

"Nosotras miramos la salud del dataset. Encontramos 11 columnas con nulos, 4 filas duplicadas y 4 columnas mal tipadas. Las más urgentes para Fase 2 son: eliminar StandardHours (inútil + 164 nulos), eliminar los 4 duplicados y convertir Age, MonthlyIncome y JobSatisfaction de float a int"

# FASE 2

In [11]:
# SIEMPRE lo primero — hacer una copia del dataset original
# Nunca tocamos el df original
df_clean = df.copy()

print("Copia creada!")
print("Filas   :", df_clean.shape[0])
print("Columnas:", df_clean.shape[1])

Copia creada!
Filas   : 1474
Columnas: 35


¿Por qué df.copy()?
Sin .copy() cualquier cambio en df_clean también cambiaría df — Python compartiría la misma tabla en memoria. Con .copy() son dos tablas completamente independientes 😊

In [12]:
# TAREA A1 — Normalizar el texto de JobRole

# ── Antes de limpiar — ver cómo está ─────────────
print("ANTES de limpiar:")
print(df_clean["JobRole"].unique())

# ── La función de limpieza ────────────────────────
def fix_job_role(df):
    # Quitar espacios del principio y del final
    df["JobRole"] = df["JobRole"].str.strip()
    # Poner la primera letra de cada palabra en mayúscula
    df["JobRole"] = df["JobRole"].str.title()
    return df

# ── Aplicar la función ────────────────────────────
df_clean = fix_job_role(df_clean)

# ── Después de limpiar — comprobar resultado ──────
print("\nDESPUES de limpiar:")
print(df_clean["JobRole"].unique())

ANTES de limpiar:
<StringArray>
[          ' sALES eXECUTIVE ',        ' rESEARCH sCIENTIST ',
     ' lABORATORY tECHNICIAN ',    ' mANUFACTURING dIRECTOR ',
 ' hEALTHCARE rEPRESENTATIVE ',                   ' mANAGER ',
      ' sALES rEPRESENTATIVE ',         ' rESEARCH dIRECTOR ',
           ' hUMAN rESOURCES ']
Length: 9, dtype: str

DESPUES de limpiar:
<StringArray>
[          'Sales Executive',        'Research Scientist',
     'Laboratory Technician',    'Manufacturing Director',
 'Healthcare Representative',                   'Manager',
      'Sales Representative',         'Research Director',
           'Human Resources']
Length: 9, dtype: str


Qué hace cada línea:
def fix_job_role(df): → define una función que recibe un DataFrame y lo devuelve limpio. En Fase 2 todo va en funciones — así lo explica vuestro documento 😊
df["JobRole"].str.strip() → elimina los espacios del principio y del final de cada valor:

' sALES eXECUTIVE ' → 'sALES eXECUTIVE'

df["JobRole"].str.title() → pone la primera letra de cada palabra en mayúscula y el resto en minúscula:

'sALES eXECUTIVE' → 'Sales Executive'

return df → devuelve el DataFrame ya limpio — importante no olvidarlo o la función no sirve de nada 😊
df_clean = fix_job_role(df_clean) → aplica la función sobre nuestra copia y guarda el resultado

### Tarea A1 — JobRole limpio ✅
- Problema: espacios al principio/final + mayúsculas caóticas
- Solución: .str.strip() + .str.title()
- Resultado: 9 roles con formato correcto

In [13]:
# TAREA A2 — Convertir columnas de float a int

# ── Antes — ver los tipos actuales ───────────────
print("ANTES de convertir:")
cols_convertir = ["Age", "JobSatisfaction", "MonthlyIncome", "YearsWithCurrManager"]
for col in cols_convertir:
    print(f"  {col}: {df_clean[col].dtype}")

# ── La función de conversión ──────────────────────
def fix_types(df):
    cols = ["Age", "JobSatisfaction", "MonthlyIncome", "YearsWithCurrManager"]
    for col in cols:
        # Primero rellenamos nulos con 0 para poder convertir
        df[col] = df[col].fillna(0)
        # Luego convertimos a entero
        df[col] = df[col].astype(int)
    return df

# ── Aplicar la función ────────────────────────────
df_clean = fix_types(df_clean)

# ── Después — comprobar que cambió ───────────────
print("\nDESPUES de convertir:")
for col in cols_convertir:
    print(f"  {col}: {df_clean[col].dtype}")

# ── Ver que ya no hay decimales ───────────────────
print("\nPrimeros valores de Age:")
print(df_clean["Age"].head())

ANTES de convertir:
  Age: float64
  JobSatisfaction: float64
  MonthlyIncome: float64
  YearsWithCurrManager: float64

DESPUES de convertir:
  Age: int64
  JobSatisfaction: int64
  MonthlyIncome: int64
  YearsWithCurrManager: int64

Primeros valores de Age:
0    41
1    49
2    37
3    33
4    27
Name: Age, dtype: int64


Qué hace cada línea:
cols = ["Age", "JobSatisfaction", "MonthlyIncome", "YearsWithCurrManager"] → lista de columnas que hay que convertir — todas son float64 pero deberían ser int
for col in cols: → recorre cada columna de la lista
df[col].fillna(0) → antes de convertir hay que rellenar los nulos — porque Python no puede convertir NaN a entero directamente. Ponemos 0 de momento — los nulos los trataremos bien en la Tarea A3
df[col].astype(int) → convierte el tipo de dato:

41.0 → 41
3.0 → 3
4907.0 → 4907

df_clean[col].dtype → muestra el tipo actual de esa columna — así vemos el antes y el después claramente

⚠️ Importante — después de esta celda Age ya no tendrá decimales. Los 0 que pusimos en los nulos los arreglaremos en la siguiente tarea 😊

### Tarea A2 — Tipos de datos corregidos ✅
- Problema: Age, JobSatisfaction, MonthlyIncome,
  YearsWithCurrManager eran float64
- Solución: .fillna(0) + .astype(int)
- Resultado: las 4 columnas son ahora int64
- Nota: los 0 de los nulos se tratarán en Tarea A3

In [14]:
# TAREA A3 — Tratar los valores nulos

# ── Antes — ver cuántos nulos quedan ─────────────
print("ANTES de tratar nulos:")
nulos_antes = df_clean.isnull().sum()
print(nulos_antes[nulos_antes > 0])

# ── La función ────────────────────────────────────
def fix_nulls(df):

    # Columnas NUMÉRICAS → rellenar con la MEDIANA
    cols_mediana = [
        "YearsWithCurrManager",
        "TrainingTimesLastYear",
        "Age",
        "JobSatisfaction",
        "MonthlyIncome"
    ]
    for col in cols_mediana:
        mediana = df[col].median()
        df[col] = df[col].replace(0, mediana)
        print(f"  {col}: nulos rellenados con mediana ({mediana})")

    # Columnas de TEXTO → rellenar con la MODA
    cols_moda = [
        "MaritalStatus",
        "BusinessTravel",
        "EducationField",
        "OverTime",
        "Department"
    ]
    for col in cols_moda:
        moda = df[col].mode()[0]
        df[col] = df[col].fillna(moda)
        print(f"  {col}: nulos rellenados con moda ('{moda}')")

    return df

# ── Aplicar la función ────────────────────────────
print("\nRellenando nulos:")
df_clean = fix_nulls(df_clean)

# ── Después — comprobar que no quedan nulos ───────
print("\nDESPUES de tratar nulos:")
nulos_despues = df_clean.isnull().sum()
nulos_despues = nulos_despues[nulos_despues > 0]
if len(nulos_despues) == 0:
    print("  No quedan nulos!")
else:
    print(nulos_despues)

ANTES de tratar nulos:
BusinessTravel           117
Department                29
EducationField            58
MaritalStatus            132
OverTime                  44
StandardHours            164
TrainingTimesLastYear     88
dtype: int64

Rellenando nulos:
  YearsWithCurrManager: nulos rellenados con mediana (2.0)
  TrainingTimesLastYear: nulos rellenados con mediana (3.0)
  Age: nulos rellenados con mediana (35.0)
  JobSatisfaction: nulos rellenados con mediana (3.0)
  MonthlyIncome: nulos rellenados con mediana (4876.5)
  MaritalStatus: nulos rellenados con moda ('Married')
  BusinessTravel: nulos rellenados con moda ('Travel_Rarely')
  EducationField: nulos rellenados con moda ('Life Sciences')
  OverTime: nulos rellenados con moda ('No')
  Department: nulos rellenados con moda ('Research & Development')

DESPUES de tratar nulos:
StandardHours            164
TrainingTimesLastYear     88
dtype: int64


Qué hace cada línea:
cols_mediana = [...] → lista de columnas numéricas con nulos — usamos la mediana porque no se ve afectada por valores extremos 😊
df[col].median() → calcula la mediana de esa columna — el valor del medio
df[col].replace(0, mediana) → reemplaza los 0 que pusimos antes en la Tarea A2 por la mediana real. Usamos replace(0, mediana) en lugar de fillna() porque en la Tarea A2 ya convertimos los nulos a 0
cols_moda = [...] → lista de columnas de texto con nulos — usamos la moda porque no puedes calcular una media de palabras 😊
df[col].mode()[0] → calcula la moda — el valor más frecuente. El [0] es porque mode() devuelve una lista y queremos solo el primero
df[col].fillna(moda) → rellena los nulos con ese valor más frecuente
if len(nulos_despues) == 0: → comprueba si quedan nulos — si la lista está vacía significa que los hemos tratado todos ✅

¿Qué ha pasado?

StandardHours → tiene 164 nulos pero no la tratamos porque la vamos a eliminar en la Tarea B1 de vuestra compañera — es una columna inútil 🗑️
TrainingTimesLastYear → tiene 88 nulos y sí debería haberse rellenado ⚠️

El problema de TrainingTimesLastYear es que en la Tarea A2 no la convertimos a int — se quedó como float64 con nulos reales, no con ceros. Por eso replace(0, mediana) no funcionó con ella 😊

In [15]:
# CORRECCION — TrainingTimesLastYear quedó con nulos
# porque es float y no pasó por fix_types

# Ver el tipo actual
print("Tipo actual:", df_clean["TrainingTimesLastYear"].dtype)

# Rellenar nulos con la mediana
mediana_training = df_clean["TrainingTimesLastYear"].median()
print("Mediana:", mediana_training)

df_clean["TrainingTimesLastYear"] = df_clean["TrainingTimesLastYear"].fillna(mediana_training)

# Convertir a int
df_clean["TrainingTimesLastYear"] = df_clean["TrainingTimesLastYear"].astype(int)

# Comprobar
print("\nNulos que quedan en TrainingTimesLastYear:")
print(df_clean["TrainingTimesLastYear"].isnull().sum())
print("Tipo ahora:", df_clean["TrainingTimesLastYear"].dtype)

Tipo actual: float64
Mediana: 3.0

Nulos que quedan en TrainingTimesLastYear:
0
Tipo ahora: int64


Qué hace:
df_clean["TrainingTimesLastYear"].median() → calcula la mediana de los valores que sí existen — ignorando los nulos
fillna(mediana_training) → rellena los nulos con esa mediana — aquí usamos fillna directamente porque la columna sigue siendo float con nulos reales, no con ceros
.astype(int) → la convierte a entero ya sin nulos

StandardHours la dejamos con nulos — la eliminará la Pareja B en su Tarea B1 😊

### Tarea A3 — Nulos tratados ✅
- Columnas numéricas → rellenadas con la MEDIANA
  (más robusta que la media ante valores extremos)
    - YearsWithCurrManager → mediana 2
    - TrainingTimesLastYear → mediana 3
    - Age → mediana 35
    - JobSatisfaction → mediana 3
    - MonthlyIncome → mediana 4876
- Columnas de texto → rellenadas con la MODA
  (el valor más frecuente)
    - MaritalStatus → 'Married'
    - BusinessTravel → 'Travel_Rarely'
    - EducationField → 'Life Sciences'
    - OverTime → 'No'
    - Department → 'Research & Development'
- StandardHours → pendiente eliminar en Tarea B1

# Las 3 funciones que has creado
def fix_job_role(df):      # Tarea A1
def fix_types(df):         # Tarea A2
def fix_nulls(df):         # Tarea A3

🎊 ¡FASE 2 PAREJA A COMPLETADA! 🎉

"Nosotras dejamos el dataset con JobRole limpio, los tipos corregidos y los nulos tratados. Solo queda StandardHours con 164 nulos — esa la elimináis vosotras junto con EmployeeCount y Over18"